internals/overlord/state/notices.go

```diff
@@ -0,0 +1,423 @@
+	sort.Slice(notices, func(i, j int) bool {
+		return notices[i].lastRepeated.Before(notices[j].lastRepeated)
+	})
+	return notices
+}
+
+// Notice return a single notice by ID, or nil if not found.
+func (s *State) Notice(id string) *Notice {
+	s.reading()
+
+	// Could use another map for lookup, but the number of notices will likely
+	// be small, and this function is probably only used rarely by the CLI, so
+	// performance is unlikely to matter.
+	for _, notice := range s.notices {
+		if notice.id == id {
+			return notice
+		}
+	}
+	return nil
+}
+
+func (s *State) flattenNotices(filter *NoticeFilter) []*Notice {
+	now := time.Now()
+	var notices []*Notice
+	for _, n := range s.notices {
+		if n.expired(now) || !filter.matches(n) {
+			continue
+		}
+		notices = append(notices, n)
+	}
+	return notices
+}
+
+func (s *State) unflattenNotices(flat []*Notice) {
+	now := time.Now()
+	s.notices = make(map[noticeKey]*Notice)
+	for _, n := range flat {
+		if n.expired(now) {
+			continue
+		}
+		uniqueKey := noticeKey{n.noticeType, n.key}
+		s.notices[uniqueKey] = n
+	}
+}
+
+// WaitNotices waits for notices that match the filter to exist or occur,
+// returning the list of matching notices ordered by the last-repeated time.
+//
+// It waits till there is at least one matching notice or the context is
+// cancelled. If there are existing notices that match the filter,
+// WaitNotices will return them immediately.
+func (s *State) WaitNotices(ctx context.Context, filter *NoticeFilter) ([]*Notice, error) {
+	s.reading()
+
+	// If there are existing notices, return them right away.
+	//
+	// State is already locked here by the caller, so notices won't be added
+	// concurrently.
+	notices := s.Notices(filter)
+	if len(notices) > 0 {
+		return notices, nil
+	}
+
+	// When the context is done/cancelled, wake up the waiters so that they
+	// can check their ctx.Err() and return if they're cancelled.
+	//
+	// TODO: replace this with context.AfterFunc once we're on Go 1.21.
+	stop := contextAfterFunc(ctx, func() {
+		// We need to acquire the cond lock here to be sure that the Broadcast
+		// below won't occur before the call to Wait, which would result in a
+		// missed signal (and deadlock).
+		s.noticeCond.L.Lock()
+		defer s.noticeCond.L.Unlock()
+
+		s.noticeCond.Broadcast()
+	})
+	defer stop()
+
+	for {
+		// Wait till a new notice occurs or a context is cancelled.
+		s.noticeCond.Wait()
+
+		// If this context is cancelled, return the error.
+		ctxErr := ctx.Err()
+		if ctxErr != nil {
+			return nil, ctxErr
+		}
+
+		// Otherwise check if there are now matching notices.
+		notices = s.Notices(filter)
+		if len(notices) > 0 {
+			return notices, nil
+		}
+	}
+}
+
+// Remove this and just use context.AfterFunc once we're on Go 1.21.
+func contextAfterFunc(ctx context.Context, f func()) func() {
+	stopCh := make(chan struct{}, 1)
+	go func() {
+		select {
+		case <-ctx.Done():
+			f()
+		case <-stopCh:
+		}
+	}()
+	stop := func() {
+		select {
```

Comment:
FWIW, this can also be just `close(stopCh)` and a no-buffer channel.

Reply:
Good call -- fixed as a drive-by in the [API PR](https://github.com/canonical/pebble/pull/296/commits/cec77391e82b54118a88db42438d7b43af6c54f5).